### PR TITLE
Tech: do not generate ESM dist for preset files

### LINKED ISSUE #21316: [Bug]: Critical dependency: require function is used in a way in which dependencies cannot be statically extracted.
### Describe the bug

After migrating, I get HMR warnings in the browser regarding 
- @storybook/react
- @storybook/addon-docs

Like so:
> ./node_modules/@storybook/addon-docs/dist/chunk-R4NKYYJA.mjs 1:37-44Critical dependency: require function is used in a way in which dependencies cannot be statically extracted./node_modules/@storybook/addon-docs/dist/chunk-R4NKYYJA.mjs 1:166-173Critical dependency: require function is used in a way in which dependencies cannot be statically extracted./node_modules/@storybook/react/dist/chunk-R4NKYYJA.mjs 1:37-44Critical dependency: require function is used in a way in which dependencies cannot be statically extracted./node_modules/@storybook/react/dist/chunk-R4NKYYJA.mjs 1:166-173Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

### To Reproduce

https://github.com/yannbf/mealdrop

### System

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ndelangen**
Closes https://github.com/storybookjs/storybook/issues/21316

## What I did

When TSup is bundling code, it will add a bit of require polyfill code to the output, when it encounters code referencing `require` in 1 of the entrypoints.

It will add the polyfill code to all entrypoints, and will refuse to treeshake it out.

So I made en excemption for preset entries, to never get a mjs file generated, because those often have references to `require` in them.

## How to test

Run the build process and check for existence of this type of code in dist:
```
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw new Error('Dynamic require of "' + x + '" is not supported');
});

export {
  __require
};
```

At least within the docs addon this no longer happens.

This code does impact all monorepo packages though, but I think that's fine. AFAIK we never use mjs exports of presets, always the cjs files.

Some day we'll convert everything over to ONLY ESM, and have all packages be type=module, but that's a project for future me.

### CODE REVIEW

**Comment by ndelangen** [comment]
before:
Screenshot 2023-05-01 at 16 38 25
Screenshot 2023-05-01 at 16 38 39


after:
Screenshot 2023-05-01 at 15 55 42

File: code/addons/docs/package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "url": "https://opencollective.com/storybook"
   },
   "license": "MIT",
+  "sideEffects": false,
```
Comment (shilman) [inline code/addons/docs/package.json:27]: Do we still need this?

Comment (ndelangen) [reply code/addons/docs/package.json:27]: No, I'll remove it.

File: scripts/prepare/bundle.ts [hunk 2]
```diff
@@ -73,11 +73,15 @@ const run = async ({ cwd, flags }: { cwd: string; flags: string[] }) => {
     optimized,
   });
 
+  const nonPresetEntries = allEntries.filter((f) => !path.parse(f).name.includes('preset'));
```
Comment (tmeasday) [inline scripts/prepare/bundle.ts:76]: Add a comment or reference to the issue as to why this is happening?

**Review by tmeasday — APPROVED** [review]
Seems like a neat solution

File: scripts/prepare/bundle.ts (hunk 2 above)
Comment (ndelangen) [reply scripts/prepare/bundle.ts:76]: good suggestion! I'll do that now.

**Pull request merged**